software/contrib/preset_manager.py

```diff
@@ -0,0 +1,191 @@
+from time import sleep
+from europi import *
+from europi_script import EuroPiScript
+
+
+class PresetManager(EuroPiScript):
+    def __init__(self):
+        super().__init__()
+        state = self.load_state_json()
+
+        self.blank_preset = ["blank preset", [0,0,0,0,0,0]]
+
+        self.presets = state.get("presets", [["preset one", [0,0,0,0,0,0]], ["preset two", [0,0,0,0,0,0]], ["preset three", [0,0,0,0,0,0]]]) # [[preset name, [preset values]], [preset name 2, [preset values 2]]
+        self.loaded_preset = state.get("loaded preset", 0)
+        self.selected_preset = k1.read_position(7) - 1
+        self.selected_letter = 0
+        self.screen = 1
+        
+        self.old_k2 = 0
+        
+        self.resolution = 128
+        
+        self.box_width = int(OLED_WIDTH/3)
+        self.box_height = int(OLED_HEIGHT/2.8)
+        
+        self.list_line_height = CHAR_HEIGHT + 4
+        
+        self.output_multiplier = MAX_OUTPUT_VOLTAGE / self.resolution
+        
+        b1.handler(self.back_button)
+        b2.handler(self.enter_button)
+        
+        self.characters = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z', '0', '1', '2', '3', '4', '5', '6', '7', '8', '9', ' ', ' ']
```

Comment:
Is that a single space twice at the end of the list of `self.characters`, or two different things that read as a single whitespace?

Is `self.characters` mutated during runtime? Is there something about MicroPython that makes that list structure more useable than a string of the same characters?

Reply:
Another simplification here would be to use the `string` package to produce your list of chars. This creates the iterable string `characters` and is probably more readable:

```
>>> import string
>>> characters = string.ascii_lowercase + string.digits + " " + " "  // Extra trailing whitespace to help knob scrolling selection.
>>> characters
'abcdefghijklmnopqrstuvwxyz0123456789  '
>>> characters[24]
'y'
```

Reply:
> Another simplification here would be to use the `string` package to produce your list of chars. This creates the iterable string `characters` and is probably more readable:
> 
> ```
> >>> import string
> >>> characters = string.ascii_lowercase + string.digits + " " + " "  # Extra trailing whitespace to help knob scrolling selection.
> >>> characters
> 'abcdefghijklmnopqrstuvwxyz0123456789  '
> >>> characters[24]
> 'y'
> ```

I never knew you could generate the alphabet that way! Would it still not be lighter weight to just manually type the alphabet as a string rather than a list though (as you've pointed out you can index it in the same way), so that the script doesn't have to do any of that conversion/concatenation on startup?

Reply:
A couple of comments.

First, I tend to believe that code can't lie about *what it does* but it cannot explain *why it does that thing*. 
So, yes, please, an explanatory comment would be great!

Second, [The Zen of Python](https://peps.python.org/pep-0020/) is really helpful to remember in making your code more idomatic Python, also known as more Pythonic. It's really useful also as a way to see how to read libraries that have been written by other people.

In this context, the general rule of "Explicit is better than implicit" is what I would apply. The most important question is, are you going to *modify the content* of `characters` during the course of the program? If yes, then a list of characters is the right way to go. Strings are IMMUTABLE.